main.go

```diff
@@ -0,0 +1,242 @@
+
+	f.SortBy = "created_at"
+	queryParams, err := query.Values(f)
+	if err != nil {
+		return nil, err
+	}
+	req, err := http.NewRequest(http.MethodGet, url, nil)
+	if err != nil {
+		return nil, err
+	}
+	req.URL.RawQuery = queryParams.Encode()
+	req.Header.Add("accept", "application/json")
+	req.Header.Add("Authorization", string(cfg.AccessToken))
+
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	var builds struct {
+		Data []build `json:"data"`
+	}
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+	if err := json.Unmarshal(body, &builds); err != nil {
+		return nil, errors.Wrap(err, string(body))
+	}
+	return builds.Data, nil
+}
+
+type filter struct {
+	Before           int    `url:"before,omitempty"`
+	Branch           string `url:"branch,omitempty"`
+	SortBy           string `url:"sort_by,omitempty"`
+	Workflow         string `url:"workflow,omitempty"`
+	PullRequestID    int    `url:"pull_request_id,omitempty"`
+	TriggerEventType string `url:"trigger_event_type,omitempty"`
+}
+
+func (f filter) String() string {
+	return fmt.Sprintf("- branch: %s\n- workflow: %s\n- pull request ID: %d\n- event type: %s", f.Branch, f.Workflow, f.PullRequestID, f.TriggerEventType)
+}
+
+type build struct {
+	Tag                     string    `json:"tag"`
+	Slug                    string    `json:"slug"`
+	Branch                  string    `json:"branch"`
+	Status                  int       `json:"status"`
+	CommitHash              string    `json:"commit_hash"`
+	StatusText              string    `json:"status_text"`
+	TriggeredAt             time.Time `json:"triggered_at"`
+	BuildNumber             int64     `json:"build_number"`
+	PullRequestID           *int64    `json:"pull_request_id"`
+	TriggeredWorkflow       string    `json:"triggered_workflow"`
+	PullRequestTargetBranch string    `json:"pull_request_target_branch"`
+}
+
+func (build build) generateFilter() filter {
+	f := filter{
+		Workflow: build.TriggeredWorkflow,
+		Branch:   build.Branch,
+		Before:   int(build.TriggeredAt.Unix()),
+	}
+	switch build.buildType() {
+	case buildTypeTag:
+		f.TriggerEventType = "tag"
+	case buildTypePullRequest:
+		f.TriggerEventType = "pull-request"
+		f.PullRequestID = int(*build.PullRequestID)
+	default: // website handles manual as push
+		f.TriggerEventType = "push"
+	}
+	return f
+}
+
+func (build build) buildType() string {
+	switch {
+	case build.Tag != "":
+		return buildTypeTag
+	case *build.PullRequestID > 0 || (build.PullRequestTargetBranch != ""):
+		return buildTypePullRequest
+	case build.CommitHash == "":
+		return buildTypeManual
+	default:
+		return buildTypePush
+	}
+}
+
+func (build build) equivalent(pair build) bool {
+	return build.Branch == pair.Branch &&
+		build.TriggeredWorkflow == pair.TriggeredWorkflow &&
+		build.buildType() == pair.buildType()
+}
+
+type builds []build
+
+func (builds builds) previous(actualBuild build) (build, error) {
```

Comment:
This function works because of the assumption, that the `builds` variable is sorted. The code that is responsible for the sorting it though lives at a very different place in the code, so, I would at least make the assumption obvious by renaming the type to `sortedBuilds`.